Create smarter cache, and use them in handlers (v2 and v3)

apelisse opened the pull request:
cc @Jefftree @alexzielenski 

I don't want to merge this yet, since I would like to try and integrate in k8s first to see if the API works well or not.

Jefftree commented on pkg/cache/cache.go:
```diff
@@ -0,0 +1,251 @@
+// [NewResultError] if the call resulted in an error.
+type Result[T any] struct {
+	data T
+	etag string
```
Can you go into detail about the format and how it should be generated/used? is it optional? I noticed in `handler.go` this etag value is ignored in some cases. Is the requirement for this etag the same as the http etag or are the requirements relaxed?
afaik with the json and proto marshalling cache, this is defined as the sha512 of the bytes. How about with the structs in the aggregator?

Edit: Okay I see that time is used as the etag in some places. This is probably not meant to be shared between threads (as mentioned in the header above). I wonder if we should use a different name to avoid potential confusion with the actual HTTP etag concept?

apelisse replied:
To be fair, this is meant as a "cache identifier", which in many occurences can be either the etag of the downloaded document, or the etag of the document that we'll serve, so it happens to often be an actual http etag. It doesn't have to, but it's a string that should correspond to the version of an object. You probably noticed that I'm sometimes generating the etag using uuid as a way to say "it's a random piece of data, I can't version it". I don't know if the name is confusing or helpful :-/ I found it useful because we know what this represents, and apart from not having to be quoted like regular http tags, they have about the same set of constraints.

lavalamp replied:
I'd call it a CacheEntry, maybe?

Jefftree commented on pkg/cache/cache.go:
```diff
@@ -0,0 +1,251 @@
+// The [NewMerger] and [NewTransformer] automatically handle
+// that for you by checking if the etag is updated before calling the
+// merging or transforming function.
+type Interface[T any] interface {
```
I feel like we can come up with a less ambiguous name

apelisse replied:
I can't find where right now, but rather than re-using the package name, you're supposed to use the name `Interface` when you would use the name of the package. So when used, it's not `cache.Cache`, but `cache.Interface` which is somewhat more useful.

lavalamp replied:
The concept implemented by this seems to be that of a handle? How about the name Handle?

lavalamp replied:
or Promise?

apelisse replied:
Updated the name (and the name of the package). this becomes `cached.Data`.

Jefftree commented on pkg/cache/cache_test.go:
```diff
@@ -0,0 +1,871 @@
+	if want := "source"; etag != want {
+		t.Fatalf("expected etag = %v, got %v", want, etag)
+	}
+	// Even if we replace with something that fails, we continue to return the success.
```
How do we surface the error in this case?

apelisse replied:
We don't, I could have a different type that doesn't do that, but it's pretty common, at least in the openapi business, that we continue to return stale data rather than new errors.

Jefftree commented on pkg/cache/cache.go:
```diff
@@ -0,0 +1,251 @@
+//
+// The caches and results are mapped by string so that associated data
+// can be retrieved.
+func NewMerger[K comparable, T, V any](mergeFn func(results map[K]Result[T]) Result[V], caches map[K]Interface[T]) Interface[V] {
```
Realistically what kind of types will we have for `K comparable?`

apelisse replied:
I only have strings so far, but it seemed odd as I was adding the generics everywhere, not to allow a different type of key? I can revert that back to string if you want.

Jefftree replied:
This is fine, I was just curious about the reasoning for the comparable, and maintaining the generic interface is a good reason.

Jefftree commented on pkg/handler3/handler.go:
```diff
@@ -92,94 +136,92 @@ func constructServerRelativeURL(gvString, etag string) string {
 
 // NewOpenAPIService builds an OpenAPIService starting with the given spec.
-func NewOpenAPIService(spec *spec.Swagger) (*OpenAPIService, error) {
+func NewOpenAPIService() *OpenAPIService {
```
As much as I want to replace this too, do we need to worry about backwards compatibility at all?

apelisse replied:
Good question, there's basically 2 things that I've had to change in k/k because of this PR, basically:
```
commit 094d2126c80978dbf1a3963828f7e0790d0dc209
Author: Antoine Pelisse <apelisse@gmail.com>
Date:   Thu Dec 29 12:12:33 2022 -0800

    Update kubernetes code for minor API changes to kube-openapi

diff --git a/staging/src/k8s.io/apiserver/pkg/server/routes/openapi.go b/staging/src/k8s.io/apiserver/pkg/server/routes/openapi.go
index e6459ce3ea2..19dbbb03443 100644
--- a/staging/src/k8s.io/apiserver/pkg/server/routes/openapi.go
+++ b/staging/src/k8s.io/apiserver/pkg/server/routes/openapi.go
@@ -56,12 +56,8 @@ func (oa OpenAPI) InstallV2(c *restful.Container, mux *mux.PathRecorderMux) (*ha
 
 // InstallV3 adds the static group/versions defined in the RegisteredWebServices to the OpenAPI v3 spec
 func (oa OpenAPI) InstallV3(c *restful.Container, mux *mux.PathRecorderMux) *handler3.OpenAPIService {
-   openAPIVersionedService, err := handler3.NewOpenAPIService(nil)
-   if err != nil {
-       klog.Fatalf("Failed to create OpenAPIService: %v", err)
-   }
-
-   err = openAPIVersionedService.RegisterOpenAPIV3VersionedService("/openapi/v3", mux)
+   openAPIVersionedService := handler3.NewOpenAPIService()
+   err := openAPIVersionedService.RegisterOpenAPIV3VersionedService("/openapi/v3", mux)
    if err != nil {
        klog.Fatalf("Failed to register versioned open api spec for root: %v", err)
    }
diff --git a/staging/src/k8s.io/kube-aggregator/pkg/controllers/openapiv3/aggregator/aggregator.go b/staging/src/k8s.io/kube-aggregator/pkg/controllers/openapiv3/aggregator/aggregator.go
index 67e793077e9..8f4edffc595 100644
--- a/staging/src/k8s.io/kube-aggregator/pkg/controllers/openapiv3/aggregator/aggregator.go
+++ b/staging/src/k8s.io/kube-aggregator/pkg/controllers/openapiv3/aggregator/aggregator.go
@@ -93,10 +93,7 @@ func BuildAndRegisterAggregator(downloader Downloader, delegationTarget server.D
        i++
    }
 
-   handler, err := handler3.NewOpenAPIService(nil)
-   if err != nil {
-       return s, err
-   }
+   handler := handler3.NewOpenAPIService()
    s.openAPIV2ConverterHandler = handler
    openAPIV2ConverterMux := mux.NewPathRecorderMux(openAPIV2Converter)
    s.openAPIV2ConverterHandler.RegisterOpenAPIV3VersionedService("/openapi/v3", openAPIV2ConverterMux)
```
It's really nothing, but yeah, do we want to keep things strictly compatible, good question ...

Jefftree replied:
If we want to backport a bug fix in kube-openapi to earlier k8s versions, will they pick up this caching feature as a byproduct?

apelisse replied:
Yeah, we would either have to branch kube-openapi which we've done a few times, or they would have to make these two minor changes 🤷

apelisse replied:
How do you feel about that?

Jefftree commented on pkg/cache/cache_test.go:
```diff
@@ -0,0 +1,871 @@
+/*
+Copyright 2022 The Kubernetes Authors.
```
nit: It's 2023 now 😮

Jefftree commented on pkg/cache/cache.go:
```diff
@@ -0,0 +1,251 @@
+	}
+	for key, oldResult := range c.cacheResults {
+		newResult, ok := results[key]
+		if !ok {
```
Is this possible given that cacheResults and results should have the same length?

apelisse replied:
Yeah, the keys can have changed. Say we cached when you had ApiService1 and ApiService2, and next time someone asks for the openapi we have ApiService1 and Apiservice3?

Jefftree replied:
Okay maybe I'm misunderstanding the code. From what I understand, the cache keys are added when the cache is created and do not change while the cache values change and are updated as needed. If the map can be modified at will, wouldn't L142 have cases where it'd fail?

```
	if len(results) != len(c.cacheResults) {
		panic(fmt.Errorf("invalid number of results: %v (expected %v)", len(results), len(c.cacheResults)))
	}
```

apelisse replied:
Yeah, you're right, the map isn't supposed to change (we re-create the merger if we want to change the map), this should panic if the key isn't present.

lavalamp commented on pkg/cache/cache.go:
```diff
@@ -0,0 +1,251 @@
+}
+
+// Result can be treated as a [Interface] if necessary.
+func (r Result[T]) Data() Result[T] {
```
What if r.err is non-nil?

apelisse replied:
I think that's solved by exporting the fields.

lavalamp commented on pkg/cache/cache.go:
```diff
@@ -0,0 +1,251 @@
+}
+
+// Split the result into its data, etag and error components.
+func (r Result[T]) Split() (T, string, error) {
```
You could also make the members exported?

apelisse replied:
Done

lavalamp commented on pkg/cache/cache.go:
```diff
@@ -0,0 +1,251 @@
+// T is the source type, V is the destination type.
+type merger[K comparable, T, V any] struct {
+	mergeFn      func(map[K]Result[T]) Result[V]
+	caches       map[K]Interface[T]
```
entries? cacheEntries?

lavalamp commented on pkg/cache/cache.go:
```diff
@@ -0,0 +1,251 @@
+}
+
+// T is the source type, V is the destination type.
+type merger[K comparable, T, V any] struct {
```
Do the merging and the caching both have to be in the same object? would it be more understandable if they were separate?

apelisse replied:
What we are caching is the result of the merge, I'm not sure how we can separate that

apelisse commented:
Updated, I've also removed all the other commits (that use the code), since I don't want to rebase everytime, just fixing the test is already pretty tedious :-)

lavalamp commented on pkg/cached/cache.go:
```diff
@@ -0,0 +1,257 @@
+//     another. It's only refreshed when the source changes.
+//   - [NewMerger]: A cache that aggregates multiple caches into one.
+//     It's only refreshed when the source changes.
+//   - [Replaceable]: A cache adapter that can ben atomically
```
```suggestion
//   - [Replaceable]: A cache adapter that can be atomically
```

lavalamp commented on pkg/cached/cache.go:
```diff
@@ -0,0 +1,257 @@
+limitations under the License.
+*/
+
+// Provides cache mechanism based on etags to lazily build, and/or cache
```
Package cache provides a ...

(godoc style)

lavalamp commented on pkg/cached/cache.go:
```diff
@@ -0,0 +1,257 @@
+}
+
+// NewResultData creates a new [Result] for a successful operation.
+func NewResultOk[T any](data T, etag string) Result[T] {
```
```suggestion
func NewResultOK[T any](data T, etag string) Result[T] {
```

lavalamp commented on pkg/cached/cache.go:
```diff
@@ -0,0 +1,257 @@
+	Err  error
+}
+
+// NewResultData creates a new [Result] for a successful operation.
```
```suggestion
// NewResultOK creates a new [Result] for a successful operation.
```

lavalamp commented on pkg/cached/cache.go:
```diff
@@ -0,0 +1,257 @@
+// dependency returned an error this time, or if the mergeFn failed
+// before, then the function is reran.
+//
+// The caches and results are mapped by string so that associated data
```
```suggestion
// The caches and results are mapped by K so that associated data
```

?

lavalamp commented on pkg/cached/cache.go:
```diff
@@ -0,0 +1,257 @@
+//
+// The caches and results are mapped by string so that associated data
+// can be retrieved.
+func NewMerger[K comparable, T, V any](mergeFn func(results map[K]Result[T]) Result[V], caches map[K]Data[T]) Data[V] {
```
is it safe for users to insert or remove map items after creating this? Maybe explain in the comment?

lavalamp commented on pkg/cached/cache.go:
```diff
@@ -0,0 +1,257 @@
+		return true
+	}
+	if len(results) != len(c.cacheResults) {
+		panic(fmt.Errorf("invalid number of results: %v (expected %v)", len(results), len(c.cacheResults)))
```
I guess this answers my question :)

Might limit reuse potential to have it not modifiable?

lavalamp replied:
ah, that makes sense. I didn't read the tests, like I said :)

Doing it this way excludes merge functions that that can save work if only some of their inputs change, I think? Might be worth mentioning that they shouldn't do that.

apelisse replied:
Yes, good point, I think I see how I can phrase that.

lavalamp commented on pkg/cached/cache.go:
```diff
@@ -0,0 +1,257 @@
+// this time, or if the transformerFn failed before, the function is
+// reran.
+func NewTransformer[T, V any](transformerFn func(Result[T]) Result[V], source Data[T]) Data[V] {
+	return NewMerger(func(caches map[string]Result[T]) Result[V] {
```
instead of string, you could make a `type transformerCacheKeyType struct{}` and be guaranteed that there is at most one entry in the map :)

lavalamp commented on pkg/cached/cache.go:
```diff
@@ -0,0 +1,257 @@
+}
+
+// NewStaticSource creates a new cache that always generates the
+// same data. This will only be called once.
```
```suggestion
// same data. This will only be called once (lazily).
```

lavalamp commented on pkg/cached/cache.go:
```diff
@@ -0,0 +1,257 @@
+// lock held). This is the type that should typically be stored in
+// structs.
+type Replaceable[T any] struct {
+	cache  atomic.Value
```
switch to `atomic.Pointer[Data[T]]`?

apelisse replied:
Oh yeah, love it!

apelisse replied:
Looks like this doesn't work with go1.18? reverted back to `atomic.Value`

lavalamp reviewed (commented):
mostly minor nits (I didn't read the tests yet)

apelisse commented:
Tests are passing, updated, thanks!

lavalamp commented on pkg/cached/cache_test.go:
```diff
@@ -0,0 +1,833 @@
+	if err := source.Get().Err; err == nil {
+		t.Fatalf("expected error, found none")
+	}
+	if count != 1 {
```
hm I thought it said errors were always re-called?

apelisse replied:
Not the `NewStaticSource` no

apelisse replied:
https://github.com/kubernetes/kube-openapi/pull/345/files#diff-a0198b5ee0e3e127b9a9b56a1fc535e84a0fc7e9fe9d789befed02641834bf00R216-R217

lavalamp commented:
/lgtm
